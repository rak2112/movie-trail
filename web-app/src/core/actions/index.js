export * from './api';
export * from './entities';
export * from './movie-details';
export * from './user';