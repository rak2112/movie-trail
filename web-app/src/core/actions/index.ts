export * from './api';
export * from './entities';
export * from './movie-details';
export * from './search-movies';
export * from './user';