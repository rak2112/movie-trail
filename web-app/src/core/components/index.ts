export * from './carousel';
export * from './form-input';
export * from './form-validation';
export * from './imageLoader';
export * from './load-error';
export * from './loader';
export * from './loader-wrapper';
export * from './movie';
export * from './movie-list';
export * from './pagination';
export * from './toast';
export * from './user-actions';

