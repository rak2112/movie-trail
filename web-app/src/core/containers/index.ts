export * from './error';
export * from './user-actions';