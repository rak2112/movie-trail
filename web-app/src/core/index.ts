// // components
// export * from './components/carousel';
// export * from './components/form-input';
// export * from './components/form-validation';
// export * from './components/imageLoader';
// export * from './components/load-error';
// export * from './components/loader';
// export * from './components/loader-wrapper';
// export * from './components/movie';
// export * from './components/movie-list';
// export * from './components/pagination';
// export * from './components/toast';
// export * from './components/user-actions';

// // containers
// export {default as Error } from './containers/error';

// // utils
// export * from './utils/movies-service';
// export * from './utils/util-service';
// export * from './utils/user-service';
// export * from './utils/localStorage';
// export * from './utils/http';

// // store 
// export * from './store/configureStore';

// // // selectors
// export * from './selectors/movie-details';
// export * from './selectors/movies';


// // interfaces 
// export * from './interfaces';

// // routes
// export * from './routes';

// // reducders
// export * from './reducers';

// // actions
// export * from './actions/api';
// export * from './actions/entities';
// export * from './actions/movie-details';
// export * from './actions/user';