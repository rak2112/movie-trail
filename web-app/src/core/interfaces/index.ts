type MovieType =
  | 'Action'
  | 'Adventure'
  | 'Animation'
  | 'Comedy'
  | 'Crime'
  | 'Documentry'
  | 'Drama'
  | 'Family'
  | 'Fantasy'
  | 'History'
  | 'Horror'
  | 'Music'
  | 'Mystery'
  | 'Romance'
  | 'Science Fiction'
  | 'TV Movie'
  | 'Thriller'
  | 'War'
  | 'Western';

export type UserMovieType= 'FAVORITE' | 'WATCHLIST';

export type UserView = 'Favorites' | 'Watchlist';

export interface Api {
  isFetching?: boolean;
  hasError?: boolean;
  errorMessage?: string;
}; 

export interface FormValues {
  name?: string;
  email: string;
  password: string;
  confirmPassword?: string;
  token?: string;
}

export interface Language {
  id: string;
  name: string;
}

export interface Movie {
  id: number;
  backdrop_path: string;
  genre_ids: number[];
  genres: Genre[];
  homepage: string;
  overview: string;
  poster_path?: string;
  runtime: string;
  release_date: string;
  spoken_languages: Language[];
  status: string;
  title: string;
  type?: string;
  vote_average: string;
};

export interface Movies {
  pageNo: number;
  totalPages: number;
  totalResults: number;
  results: Movie[];
  redirection? : boolean;
};

export interface MoviesData {
  pageNo: number;
  total_results: number;
  total_pages: number;
  results?: Movie[];
};

export interface MoviesRes {
  movies: Movie[]
};

export interface Genre {
  id: number;
  name: MovieType;
};

export interface GenresMap {
  [id: number]: Genre;
};

export interface UserMovie extends Movie {
  _id?: string;
  movieType?: UserMovieType;
}

export interface User {
  email?: string;
  displayName?: string;
  loggedIn?: boolean;
  movies?: UserMovie[];
  resetMessage?: string;
  token?: string;
}

export interface UserMovies {
  [id: string]: UserMovie;
}
export interface UserMovieMap {
  favorites: UserMovies;
  watchlist: UserMovies;
}

export interface Cast {
  character: string;
  id: string;
  name: string;
  profile_path?: string;
}

export interface Crew extends Cast {
  department: string;
  job: string;
}

export interface Poster {
  file_path: string;
}

export interface Persons {
  id?: string;
  cast: Cast[];
  crew: Crew[];
}

export interface Posters {
  id?: string;
  backdrops: Poster[];
  posters: Poster[];
}

export interface Video {
  id: string;
  key: string;
  name: string;
}

export interface Videos {
  id: string;
  results: Video[]
}

export interface MovieDetail {
  isFetching?: boolean;
  hasError?: boolean;
  details: Movie;
  images: Posters;
  persons: Persons;
  videos: Videos
}