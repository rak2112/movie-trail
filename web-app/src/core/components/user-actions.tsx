import React from 'react';
import styled from 'react-emotion';

import { Movie, UserMovieMap, UserView } from '../interfaces';

const Container = styled('div')`
  /* padding: 3px 1px 5px; */
  .user-actions {
    text-align: end;
    background: #000;
    font-size: 22px;
    opacity: 0.9;
    text-align: right !important;
  }
  a {
    color: #cc3300;
    margin-right: 14px;
    cursor: pointer;
  }
  span.glyphicon-heart {
    font-size: 25px;
  }
  span.add-to-list {
      color: #fff;
  }
  span.rating {
    color: #cc0;
    padding-left: 10px;
    float: left;
    span {
      float: right;
      padding-left: 3px;
      top: 3px;
    }
  }
  span.year {
    padding-right: 102px;
  }
  span.glyphicon-calendar {
    padding-right: 7px;
  }
`;

export interface UserAction {
  movie: Movie;
  userMovies: UserMovieMap
  userView?: UserView;
  addMovie: (movie: Movie) => void;
  deleteMovie: (movie: Movie) => void;
}

const getYear = (date: string) => new Date(date).getFullYear();

export const UserActions = ({addMovie, deleteMovie, movie, userView, userMovies}: UserAction) => {
  return (
    <Container>
     <div className="user-actions">
      <span className='rating'>{movie.vote_average}<span className="glyphicon glyphicon-star"/></span>
      { movie.release_date &&
        <span className='year'><span className="glyphicon glyphicon-calendar"/>{getYear(movie.release_date)}</span>
      }
      { userMovies.favorites[movie.id] &&
        <a title="Remove From Favorites"
          className="remove-from-favorites"
          onClick={deleteMovie.bind(UserActions, {...movie, id: userMovies.favorites[movie.id]._id})}>
          <span className="glyphicon glyphicon-heart"/>
        </a>
      }  
      { (!userMovies.favorites[movie.id] && userView !== 'Watchlist') &&
        <a title="Add to Favorites"
          className="add-to-favorites"
          onClick={addMovie.bind(UserActions, {...movie, movieType: 'FAVORITE'})}>
          <span className="glyphicon glyphicon-heart add-to-list"/>
        </a>
      }

      {
        userMovies.watchlist[movie.id] &&
        <a title="Remove From List"
          className="remove-from-watchlist"
          onClick={deleteMovie.bind(UserActions, {...movie, id: userMovies.watchlist[movie.id]._id})}>
          <span className="glyphicon glyphicon-bookmark"/>
        </a>
      }
      { (!userMovies.watchlist[movie.id] && userView !== 'Favorites') &&
        <a title="Add to List"
          className="add-to-watchlist"
          onClick={addMovie.bind(UserActions, {...movie, movieType: 'WATCHLIST'})}>
          <span className="glyphicon glyphicon-bookmark add-to-list"/>
        </a>
      }
     </div>
    </Container>
  )
};