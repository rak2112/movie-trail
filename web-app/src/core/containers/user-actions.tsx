import React, { memo, useCallback, useEffect, useMemo, useState } from 'react';
import styled from 'react-emotion';
import { connect, MapStateToProps } from 'react-redux';

import { find } from 'lodash';
import { addMovie, deleteMovie } from '../../core/actions';
import { AddRemoveFavorite } from '../../core/components/addRemoveFavs';
import { AddRemoveWatchList } from '../../core/components/addRemoveWatchList';
import { State } from '../../core/reducers/index';
import { getUserActionProps } from '../../core/selectors/index';
import { GenresMap, Movie, User, UserMovie, UserMovieMap,  } from '../interfaces';

const Container = styled('div')`
  /* padding: 3px 1px 5px; */
  .user-actions {
    text-align: end;
    background: #000;
    font-size: 22px;
    opacity: 0.9;
    text-align: right !important;
  }
  a {
    color: #cc3300;
    margin-right: 14px;
    cursor: pointer;
  }
  span.glyphicon-heart {
    font-size: 25px;
  }
  span.add-to-list {
      color: #fff;
  }
  span.rating {
    color: #cc0;
    padding-left: 10px;
    float: left;
    span {
      float: right;
      padding-left: 3px;
      top: 3px;
    }
  }
  span.year {
    padding-right: 102px;
  }
  span.glyphicon-calendar {
    padding-right: 7px;
  }
`;

export interface UserAction {
  movie: Movie;
  userView?: string; 
  profileView?: boolean;
}

export interface StateProps {
  favorites: UserMovie[];
  user?: User;
  watchList: UserMovie[];
}

interface DispatchProps {
  addMovie: (movie: UserMovie) => void;
  deleteMovie: (movie: UserMovie) => void;
}
export type Actions = UserAction & StateProps & DispatchProps;

const getYear = (date: string) => new Date(date).getFullYear();

export const UserActions = ({profileView, user, userView, movie, favorites, watchList, ...props }: Actions) => {

  const [favorite, setFavorite] = useState(!!find(favorites, ['id', movie.id]));
  const [inWatch, setWatchList] = useState(!!find(watchList, ['id', movie.id]));

  const addRemoveFavorites = useCallback(() => {
    setFavorite((fav: boolean) => !fav);
    
    if (!favorite) {
      props.addMovie({...movie, movieType: 'FAVORITE'});
    }
    else {
      const {_id: id}: any = favorites.find(({id: movieId}: any) => movieId === movie.id);
      props.deleteMovie({...movie, id });
    }
  }, [favorite, favorites]);

  const addRemoveWatchList = useCallback(() => {
    setWatchList((inList: boolean) => !inList);
    if (!inWatch) {
      props.addMovie({...movie, movieType: 'WATCHLIST'});
    }
    else {
      const {_id: id}: any = watchList.find(({id: movieId}: any) => movieId === movie.id);
      props.deleteMovie({...movie, id });
    }
  }, [inWatch, watchList]);

  return (
    <Container>
      <div className="user-actions">
        <span className='rating'>{movie.vote_average}<span className="glyphicon glyphicon-star"/></span>
        { movie.release_date &&
          <span className='year'><span className="glyphicon glyphicon-calendar"/>{getYear(movie.release_date)}</span>
        }

        { (user && !profileView) ? 
          <>
            <AddRemoveFavorite favorite={favorite} onClick={addRemoveFavorites}/>
            <AddRemoveWatchList inWatch={inWatch} onClick={addRemoveWatchList}/>
          </>
          : null
        }

        {
          (profileView && userView === 'Favorites') ?
          <AddRemoveFavorite favorite={favorite} onClick={addRemoveFavorites}/> : null
        }

        {
          (profileView && userView === 'Watchlist') ?
          <AddRemoveWatchList inWatch={inWatch} onClick={addRemoveWatchList}/> : null
        }
      </div>
    </Container>
  )
};

const mapStateToProps: MapStateToProps<any, any, any> = (state: State) => {
  const props = getUserActionProps(state);
  return { ...props };
};

export default connect<StateProps, DispatchProps>(mapStateToProps, {
  addMovie,
  deleteMovie
}) (UserActions);