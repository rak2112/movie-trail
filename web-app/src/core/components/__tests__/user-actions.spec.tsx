import { shallow } from 'enzyme';
import React from "react";
import { create } from 'react-test-renderer';
import { UserMovieType } from 'src/core/interfaces';

import { UserAction, UserActions } from '../user-actions';


describe(`Carousel Component`, () => {

  const defaultProps = {
    addMovie: jest.fn(),
    deleteMovie: jest.fn(),
    movie: {
      id: 2,
      backdrop_path: 'path',
      genre_ids: [1, 2],
      genres: [],
      homepage: 'homepage',
      overview: 'overview',
      runtime: '107',
      release_date: '2-04-19',
      spoken_languages: [{id: '2', name: 'English'}],
      status: 'status',
      title: 'title',
      vote_average: '8'
    },
    userMovies: {
      favorites: {},
      watchlist: {}
    }
  } as UserAction;

  const propsWithList = {
    ...defaultProps,
    userMovies: {
      favorites: {
        ...defaultProps.userMovies.favorites,
        2: {
          ...defaultProps.movie, 
          _id: ':Id', 
          movieType: 'FAVORITE' as UserMovieType
        }
      },
      watchlist: {
        2: {
          ...defaultProps.movie,
          _id: ':Id',
          movieType: 'WATCHLIST' as UserMovieType
        }
      }
    }
  };

  it(`should render when movie is already in user favorite list`, () => {
    const props = {
      ...defaultProps,
      userMovies: {
        favorites: {
          ...defaultProps.userMovies.favorites,
          2: {
            ...defaultProps.movie, 
            _id: 'someId', 
            movieType: 'FAVORITE' as UserMovieType
          }
        },
        watchlist: {}
      }
    };

    const wrapper = create(
      <UserActions {...props}/>
    ).toJSON();
    
    expect(wrapper).toMatchSnapshot();
  });

  it(`should render when movie is not in user list`, () => {
    const wrapper = create(
      <UserActions {...defaultProps}/>
    ).toJSON();
    
    expect(wrapper).toMatchSnapshot();
  });

  it(`should render and should call addMovie with favorite`, () => {
    const wrapper = shallow(
      <UserActions {...defaultProps}/>
    );

    const addToFav = wrapper.find('.add-to-favorites');
    
    expect(wrapper.find('.glyphicon-star').length).toBe(1);
    expect(addToFav.prop('title')).toContain('Add to Favorites');
    expect(wrapper.find('.add-to-watchlist').prop('title')).toBe('Add to List');

    addToFav.simulate('click');
    expect(defaultProps.addMovie).toBeCalledWith({
      id: 2,
      backdrop_path: 'path',
      genre_ids: [1, 2],
      genres: [],
      homepage: 'homepage',
      movieType: 'FAVORITE',
      overview: 'overview',
      runtime: '107',
      release_date: '2-04-19',
      spoken_languages: [{id: '2', name: 'English'}],
      status: 'status',
      title: 'title',
      vote_average: '8'
    });
  });

  it(`should render and should call addMovie with watchlist`, () => {
    const wrapper = shallow(
      <UserActions {...defaultProps}/>
    );

    const addToList = wrapper.find('.add-to-watchlist');
    
    expect(wrapper.find('.glyphicon-star').length).toBe(1);
    expect(addToList.prop('title')).toBe('Add to List');

    addToList.simulate('click');
    expect(defaultProps.addMovie).toBeCalledWith({
      id: 2,
      backdrop_path: 'path',
      genre_ids: [1, 2],
      genres: [],
      homepage: 'homepage',
      movieType: 'WATCHLIST',
      overview: 'overview',
      runtime: '107',
      release_date: '2-04-19',
      spoken_languages: [{id: '2', name: 'English'}],
      status: 'status',
      title: 'title',
      vote_average: '8'
    });
  });

  it(`should render when movie is already in user favorite and watchlist`, () => {
    
    const wrapper = shallow(
      <UserActions {...propsWithList}/>
    );

    const removeFromFavs = wrapper.find('.remove-from-favorites');
    const removeFromList = wrapper.find('.remove-from-watchlist');
    

    expect(wrapper.find('.add-to-favorites').length).toBe(0);
    expect(removeFromFavs.prop('title')).toContain('Remove From Favorites');

    removeFromFavs.simulate('click');
    expect(defaultProps.deleteMovie).toHaveBeenCalled();
  
    expect(wrapper.find('.add-to-watchlist').length).toBe(0);
    expect(removeFromList.prop('title')).toContain('Remove From List');
    
    removeFromList.simulate('click');
    expect(defaultProps.deleteMovie).toHaveBeenCalledWith({
      id: ':Id',
      backdrop_path: 'path',
      genre_ids: [1, 2],
      genres: [],
      homepage: 'homepage',
      overview: 'overview',
      runtime: '107',
      release_date: '2-04-19',
      spoken_languages: [{id: '2', name: 'English'}],
      status: 'status',
      title: 'title',
      vote_average: '8'
    });

  });

});