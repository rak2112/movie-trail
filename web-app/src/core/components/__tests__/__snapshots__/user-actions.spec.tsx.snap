// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Component should render when movie is already in user favorite list 1`] = `
<div
  className="css-1lw9bxc"
>
  <div
    className="user-actions"
  >
    <span
      className="rating"
    >
      8
      <span
        className="glyphicon glyphicon-star"
      />
    </span>
    <a
      className="remove-from-favorites"
      onClick={[Function]}
      title="Remove From Favorites"
    >
      <span
        className="glyphicon glyphicon-heart"
      />
    </a>
    <a
      className="add-to-watchlist"
      onClick={[Function]}
      title="Add to List"
    >
      <span
        className="glyphicon glyphicon-bookmark add-to-list"
      />
    </a>
  </div>
</div>
`;

exports[`Carousel Component should render when movie is not in user list 1`] = `
<div
  className="css-1lw9bxc"
>
  <div
    className="user-actions"
  >
    <span
      className="rating"
    >
      8
      <span
        className="glyphicon glyphicon-star"
      />
    </span>
    <a
      className="add-to-favorites"
      onClick={[Function]}
      title="Add to Favorites"
    >
      <span
        className="glyphicon glyphicon-heart add-to-list"
      />
    </a>
    <a
      className="add-to-watchlist"
      onClick={[Function]}
      title="Add to List"
    >
      <span
        className="glyphicon glyphicon-bookmark add-to-list"
      />
    </a>
  </div>
</div>
`;
